### PR TITLE
[Move] Implement Flower Shield

### PR DESCRIPTION
**Pull request by torranx**
## What are the changes?
- implemented [flower shield](https://bulbapedia.bulbagarden.net/wiki/Flower_Shield_(move))

## Why am I doing these changes?
- unimplemented move

## What did change?
- implemented flower shield

### Screenshots/Videos

https://github.com/pagefaultgames/pokerogue/assets/68144167/7de33be2-509b-4bcc-a005-78c2e00be661

Screenshot 2024-06-23 at 11 46 16 PM

### How to test the changes?
- give your mon Flower Shield
- have a grass type mon in field
- `$ npm run test flower_shield`

## Checklist
- [x] There is no overlap with another PR?
- [x] The PR is self-contained and cannot be split into smaller PRs?
- [x] Have I provided a clear explanation of the changes?
- [x] Have I tested the changes (manually)?
    - [x] Are all unit tests still passing? (`npm run test`)
- [ ] Are the changes visual?
  - [x] Have I provided screenshots/videos of the changes?

### CODE REVIEW

File: src/data/move.ts [hunk 1]
```diff
@@ -7238,7 +7238,7 @@ export function initMoves() {
       .attr(AddArenaTagAttr, ArenaTagType.CRAFTY_SHIELD, 1, true, true),
     new StatusMove(Moves.FLOWER_SHIELD, Type.FAIRY, -1, 10, -1, 0, 6)
       .target(MoveTarget.ALL)
-      .unimplemented(),
+      .attr(StatChangeAttr, BattleStat.DEF, 1, false, (user, target, move) => target.getTypes().includes(Type.GRASS) && !target.getTag(BattlerTagType.FLYING) && !target.getTag(BattlerTagType.UNDERGROUND)),
```
Comment (Tempo-anon) [inline src/data/move.ts:7241]: I think pokemon using dive or in the shadow realm should also be excluded

File: src/field/pokemon.ts [hunk 2]
```diff
@@ -286,6 +286,26 @@ export default abstract class Pokemon extends Phaser.GameObjects.Container {
     return this.isAllowedInBattle() && !!this.scene && (!onField || this.isOnField());
   }
 
+  /**
+   * Checks if the Pokémon is in a semi-invulnerable state due to certain tags.
+   *
+   * @returns `true` if the Pokémon is in a semi-invulnerable state, otherwise `false`.
+   */
+  isInSemiVulnerableState(): boolean {
```
Comment (innerthunder) [inline src/field/pokemon.ts:294]: A potentially cleaner alternative to this function:

```ts
!!this.getTag(HideSpriteTag);
```

It doesn't include the check that `this.scene` is defined, but `getTag()` has a null check for `this.summonData` that might accomplish the same thing.

Comment (torranx) [reply src/field/pokemon.ts:294]: i agree this is cleaner, but would this be intuitive to use if it is not wrapped inside a function?

another option would be to use named import instead just to be more descriptive ie` import HideSpriteTag as SemiInvulnerableTag`

Comment (innerthunder) [reply src/field/pokemon.ts:294]: I think with a named import it would be clear enough.

Comment (torranx) [reply src/field/pokemon.ts]: done, thanks!

**Review by innerthunder — APPROVED** [review]

File: src/data/move.ts [hunk 3]
```diff
@@ -1,7 +1,7 @@
 import { ChargeAnim, MoveChargeAnim, initMoveAnim, loadMoveAnimAssets } from "./battle-anims";
 import { BattleEndPhase, MovePhase, NewBattlePhase, PartyStatusCurePhase, PokemonHealPhase, StatChangePhase, SwitchSummonPhase } from "../phases";
 import { BattleStat, getBattleStatName } from "./battle-stat";
-import { EncoreTag } from "./battler-tags";
+import { EncoreTag, HideSpriteTag as SemiInvulnerableTag } from "./battler-tags";
```
Comment (Tempo-anon) [inline src/data/move.ts:4]: Man... we really need to have some consistency in naming. `HideSpriteTag` is also imported in `phases.ts` but there it's been renamed as `HiddenTag`

Comment (torranx) [reply src/data/move.ts]: I've renamed all instances of `HideSpriteTag` to `SemiInvulnerableTag` instead which makes a lot more sense

**Review by Tempo-anon — APPROVED** [review]

**Pull request merged**